pkg/storage/snapshot/restore_test.go

```diff
@@ -936,6 +950,235 @@ var _ = Describe("Restore controller", func() {
+				preferenceObj          *instancetypev1alpha2.VirtualMachinePreference
+				preferenceSnapshotCR   *appsv1.ControllerRevision
+				preferenceOriginalCR   *appsv1.ControllerRevision
+			)
+
+			expectCreateVM := func(vm *v1.VirtualMachine) {
+				newVMUID := vm.UID
+				vm.UID = ""
+				vm.ResourceVersion = ""
+				vm.Annotations = map[string]string{"restore.kubevirt.io/lastRestoreUID": "restore-uid"}
+				vmInterface.EXPECT().
+					Create(vm).
+					Do(func(objs ...interface{}) {
+						vm.UID = newVMUID
+					}).Return(vm, nil)
+			}
+
+			expectUpdateVMRestored := func(vm *v1.VirtualMachine) {
+				expectedUpdatedVM := vm.DeepCopy()
+				expectedUpdatedVM.Annotations = map[string]string{"restore.kubevirt.io/lastRestoreUID": "restore-uid"}
+				vmInterface.EXPECT().
+					Update(expectedUpdatedVM).
+					Do(func(objs ...interface{}) {
+						updatedVM := objs[0].(*v1.VirtualMachine)
+						Expect(*updatedVM).To(Equal(*expectedUpdatedVM))
+					}).Return(expectedUpdatedVM, nil)
+			}
+
+			expectUpdateVMRestoreUpdatingTargetSpec := func(vmRestore *snapshotv1.VirtualMachineRestore) {
+				expectedUpdatedRestore := vmRestore.DeepCopy()
+				expectedUpdatedRestore.ResourceVersion = "1"
+				expectedUpdatedRestore.Status.Conditions = []snapshotv1.Condition{
+					newProgressingCondition(corev1.ConditionTrue, "Updating target spec"),
+					newReadyCondition(corev1.ConditionFalse, "Waiting for target update"),
+				}
+				expectVMRestoreUpdate(kubevirtClient, expectedUpdatedRestore)
+			}
+
+			BeforeEach(func() {
+				virtClient.EXPECT().AppsV1().Return(k8sClient.AppsV1()).AnyTimes()
+
+				originalVM = createSnapshotVM()
+				restore = createRestoreWithOwner()
+
+				vmSnapshot = createSnapshot()
+				vmSnapshotContent = createVirtualMachineSnapshotContent(vmSnapshot, originalVM, nil)
+				vmSnapshotContent.Status = &snapshotv1.VirtualMachineSnapshotContentStatus{
+					CreationTime: timeFunc(),
+					ReadyToUse:   &t,
+				}
+
+				vmSnapshot.Status.VirtualMachineSnapshotContentName = &vmSnapshotContent.Name
+				vmSnapshotSource.Add(vmSnapshot)
+
+				instancetypeObj = createInstancetype()
+				var err error
+				instancetypeOriginalCR, err = instancetype.CreateControllerRevision(originalVM, instancetypeObj)
+				Expect(err).ToNot(HaveOccurred())
+				crSource.Add(instancetypeOriginalCR)
+
+				instancetypeSnapshotCR = createInstancetypeVirtualMachineSnapshotCR(originalVM, vmSnapshot, instancetypeObj)
+				crSource.Add(instancetypeSnapshotCR)
+
+				preferenceObj = createPreference()
+				preferenceOriginalCR, err = instancetype.CreateControllerRevision(originalVM, preferenceObj)
+				Expect(err).ToNot(HaveOccurred())
+				crSource.Add(preferenceOriginalCR)
+
+				preferenceSnapshotCR = createInstancetypeVirtualMachineSnapshotCR(originalVM, vmSnapshot, preferenceObj)
+				crSource.Add(preferenceSnapshotCR)
+
+			})
+
+			It("should restore instancetype ControllerRevision into existing VirtualMachine", func() {
```

Comment:
Can you make tables of the instancetype/preference tests? They look very similar.

Reply:
ACK done.